core/src/test/java/apoc/trigger/TriggerClusterRoutingTest.java

```diff
@@ -0,0 +1,71 @@
+import apoc.util.TestcontainersCausalCluster;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.neo4j.driver.types.Node;
+import org.neo4j.internal.helpers.collection.MapUtil;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static apoc.trigger.TriggerDeprecatedProcedures.SYS_NON_LEADER_ERROR;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class TriggerClusterRoutingTest {
+
+    private static TestcontainersCausalCluster cluster;
+
+    @BeforeClass
+    public static void setupCluster() {
+        cluster = TestContainerUtil
+                .createEnterpriseCluster(3, 1, Collections.emptyMap(), MapUtil.stringMap(
+                        "NEO4J_dbms_routing_enabled", "true",
+                        "apoc.trigger.enabled", "true"
+                ));
+    }
+
+    @AfterClass
+    public static void bringDownCluster() {
+        if (cluster != null) {
+            cluster.close();
+        }
+    }
+
+    @Test
+    public void testTriggerAddAllowedOnlyInLeaderMember() {
+        final List<Neo4jContainerExtension> members = cluster.getClusterMembers();
+        assertEquals(4, members.size());
+        for (Neo4jContainerExtension container: members){
+            // we skip READ_REPLICA members
+            final String readReplica = TestcontainersCausalCluster.ClusterInstanceType.READ_REPLICA.toString();
+            if (readReplica.equals(container.getEnvMap().get("NEO4J_dbms_mode")) || container.getSession() == null) {
+                continue;
+            }
+            final String systemRole = TestContainerUtil.singleResultFirstColumn(container.getSession(), "CALL dbms.cluster.role('system')");
+            if ("LEADER".equals(systemRole)) {
+                container.getSession().run("CALL apoc.trigger.add($name, 'UNWIND $createdNodes AS n SET n.ts = timestamp()',{})",
+                        Map.of("name", "trigger-" + container.getContainerName()));
+
+                container.getSession().run("CREATE (f:Foo)");
+                TestContainerUtil.testCall(container.getSession(), "MATCH (f:Foo) RETURN f",
+                        (row) -> assertTrue(((Node) row.get("f")).containsKey("ts")));
+            } else {
+                try {
+                    TestContainerUtil.testCall(container.getSession(), "CALL apoc.trigger.add($name, 'UNWIND $createdNodes AS n SET n.ts = timestamp()',{})",
+                            Map.of("name", "trigger-" + container.getContainerName()),
+                            row -> fail("Should fail because of non leader trigger addition"));
+                } catch (RuntimeException e) {
+                    String errorMsg = e.getMessage();
+                    assertTrue("The actual message is: " + errorMsg, errorMsg.contains(SYS_NON_LEADER_ERROR));
+                }
+            }
+        }
+    }
+}
```

Comment:
I would like to encourage writing these tests the way a user would write his queries in production using the Java Driver. This is because we can be sure that our logic works as expected for our users, and they can also be used as an example code.

Users' application code won't have access to the container information of their deployment. Instead they have to open a session against "system" which will then be routed by the client (if client-side routing is enabled) or by the DBMS (if server-side routing is enabled). The routing is transparent to the user.

Then the user would have to open another session against his "neo4j" database to create nodes, etc. This should also be automatically routed by the drivers.

Reply:
As below, I think it is better to wait for the annotations `@SystemOnlyProcedure` for "system" session's tests

Reply:
I think what @AzuObs meant is that we shouldn't be using `container.getSession())` in these tests because a client code normally doesn't have access to the `container` data structure. He meant opening a session with bolt I think

Reply:
I'm happy to merge the PR as is and then we can iterate in some bits that can be improved